### PR TITLE
maptm: Upstream recent fixes in MAP-T manager

### PR DESCRIPTION
**Pull request by MedHediDrd**
This commit contains the new changes applied to MAP-T manager.

### CODE REVIEW

**Comment by MalekSST** [comment]
###############################################################
Ticket: WROUTSAGE-1214 - Sometimes TCP port is not from the MAP port set
###############################################################
Root cause: The kernel module is not cleaning the port map when
changing configuration.

Solution: When reconfiguring MAP-T, clean the port mapping and stop
MAP-T kernel module.

Test scenario:
1- Start multiple iperf sessions using the same source por.
2- Change WAN server configurations.
3- Start a new iperf session with the same source port and destination
to the same server.

Test results: Ok

###############################################################
Ticket: WROUTSAGE-1253 - Router told to switch from MAP-T to
Dual-Stack. Node_State table still shows MAP-T
###############################################################
Root cause: When the cloud sets the MAP-T support value to false, the
eligibility state machine is not restarting correctly.

Solution: When the cloud inserts a new entry in Node_Config table,
restart the eligibility state machine based on the support value.

Test scenario:
1- When the gateway is up on the NOC, check Node_Config MAP-T entry
is inserted.
2- Based on the support value, check that the WAN mode in Node_State
table is set correctly.
3- Check that the gateway is getting the correct address(es) based on
its posture.

Test results: Ok

###############################################################
Ticket: WROUTSAGE-1191 - MAP-T only works with 192.168.1.1/24 LAN subnet
###############################################################
Root cause: Subnetcidr4 configuration is static.

Solution: Make subnetcidr4 configuration dynamic.

Test scenario:
1- Change the BR_LAN subnet.
2- Check that MAP-T is configured with the new subnet.

Test results: Ok

File: src/maptm/src/maptm_config.c [hunk 1]
```diff
@@ -441,6 +442,36 @@ bool stop_mapt(void)
...
+            SCHEMA_COLUMN(Wifi_Inet_Config, if_name),
+            MAPT_IFC_LAN,
+            &iconf);
+
+    if (!ret)
+    {
+        LOGE("%s: Failed to get Interface config", __func__);
+        return NULL;
+    }
+    if(!osn_ip_addr_from_str(&netmask, iconf.netmask))
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_config.c:465]:
```suggestion
    if (!osn_ip_addr_from_str(&netmask, iconf.netmask))
```

File: src/maptm/src/maptm_config.c [hunk 2]
```diff
@@ -450,8 +481,14 @@ bool config_mapt(void)
         return false;
     }
 
-    /* Hard Coded value: Must be changed */
-    char subnetcidr4[20] = "192.168.1.1/24";
+    // Get LAN Subnet
+    char subnetcidr4[20] = "";
+    set_subnetcidr4(subnetcidr4);
+    if(subnetcidr4 == NULL)
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_config.c:487]:
```suggestion
    if (subnetcidr4 == NULL)
```

File: src/maptm/src/maptm_eligibility.c [hunk 3]
```diff
@@ -203,13 +204,23 @@ static void callback_DHCP_Option(
         case OVSDB_UPDATE_NEW:
             if (((new->tag) == 26) && (new->enable) && !strcmp((new->type), "rx"))
             {
+                if(strcmp(new->value, strucWanConfig.iapd) && !strcmp("MAP-T", strucWanConfig.mapt_mode))
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:207]:
```suggestion
                if (strcmp(new->value, strucWanConfig.iapd) && !strcmp("MAP-T", strucWanConfig.mapt_mode))
```

File: src/maptm/src/maptm_eligibility.c [hunk 4]
```diff
@@ -203,13 +204,23 @@ static void callback_DHCP_Option(
...
+                {
+                    LOGD("%s IAPD is changed, we need to update MAP-T configuration", __func__);
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:209]:
```suggestion
                    LOGD("%s: IAPD is changed, we need to update MAP-T configuration", __func__);
```

File: src/maptm/src/maptm_eligibility.c [hunk 5]
```diff
@@ -203,13 +204,23 @@ static void callback_DHCP_Option(
...
+                    needReconfigure = true;
+                }
                 snprintf(strucWanConfig.iapd, sizeof(strucWanConfig.iapd), "%s", new->value);
             }
 
             if ((new->tag) == 95)
             {
                 if (new->enable)
                 {
+                    if(strcmp(new->value, strucWanConfig.mapt_95_value))
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:219]:
```suggestion
                    if (strcmp(new->value, strucWanConfig.mapt_95_value))
```

File: src/maptm/src/maptm_eligibility.c [hunk 6]
```diff
@@ -203,13 +204,23 @@ static void callback_DHCP_Option(
...
+                    {
+                        LOGD("%s Option 95 is changed, we need to update MAP-T configuration", __func__);
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:221]:
```suggestion
                        LOGD("%s: Option 95 is changed, we need to update MAP-T configuration", __func__);
```

File: src/maptm/src/maptm_eligibility.c [hunk 7]
```diff
@@ -221,7 +232,7 @@ static void callback_DHCP_Option(
                         maptm_wan_mode();
                     }
                     else if (strucWanConfig.link_up
-                             && (!strcmp("Dual-Stack", strucWanConfig.mapt_mode)))
+                            && (!strcmp("Dual-Stack", strucWanConfig.mapt_mode)))
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:235]:
```suggestion
                             && (!strcmp("Dual-Stack", strucWanConfig.mapt_mode)))
```

File: src/maptm/src/maptm_eligibility.c [hunk 8]
```diff
@@ -318,12 +331,19 @@ static void callback_DHCP_Option(
             if ((old->tag) == 26)
             {
                 char *flag = NULL;
-                if (!strcmp(old->value,strucWanConfig.iapd)) strucWanConfig.iapd[0] = '\0';
+                if (!strcmp(old->value, strucWanConfig.iapd)) strucWanConfig.iapd[0] = '\0';
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_eligibility.c:334]:
```suggestion
                if (!strcmp(old->value, strucWanConfig.iapd))
                {
                    strucWanConfig.iapd[0] = '\0';
                }
```

File: src/maptm/src/maptm_config.c [hunk 9]
```diff
@@ -450,8 +481,14 @@ bool config_mapt(void)
...
     }
 
-    /* Hard Coded value: Must be changed */
-    char subnetcidr4[20] = "192.168.1.1/24";
+    // Get LAN Subnet
+    char subnetcidr4[20] = "";
+    set_subnetcidr4(subnetcidr4);
+    if (subnetcidr4 == NULL)
+    {
+        LOGE("Unable to configure MAP-T option Unable to get subnetcidr4");
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_config.c:489]:
```suggestion
        LOGE("Unable to get subnetcidr4");
```

or

```suggestion
        LOGE("Unable to configure MAP-T option (unable to get subnetcidr4)");
```

whichever you think is more appropriate

File: src/maptm/src/maptm_config.c [hunk 10]
```diff
@@ -441,6 +442,36 @@ bool stop_mapt(void)
     return true;
 }
 
+// Get LAN Subnet
+char* set_subnetcidr4(char *subnetcidr4)
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_config.c:446]:
```suggestion
char* get_subnetcidr4(char *subnetcidr4)
```

File: src/maptm/src/maptm_config.c [hunk 11]
```diff
@@ -450,8 +481,14 @@ bool config_mapt(void)
         return false;
     }
 
-    /* Hard Coded value: Must be changed */
-    char subnetcidr4[20] = "192.168.1.1/24";
+    // Get LAN Subnet
+    char subnetcidr4[20] = "";
+    set_subnetcidr4(subnetcidr4);
+    if (subnetcidr4 == NULL)
```
Comment (lkristan-plume) [inline src/maptm/src/maptm_config.c:487]: `subnetcidr4` cannot be NULL, it's on stack, so check the return value:

```suggestion
    if (get_subnetcidr4(subnetcidr4) == NULL)
```

Also changed `set_subnetcidr4` to `get_subnetcidr4` (in line with the suggestion above).

**Pull request merged**